### ISSUE TITLE
clusterInstanceHostPattern should be optional for custom host names

### ISSUE THREAD

**Issue by tomchiverton** [awslabs/aws-mysql-jdbc#296]
### Describe the feature

When a custom host name is used as the host in the JDBC URL, then opaquely named clusterInstanceHostPattern paramater must be added. This should have a sensible default instead. 

### Use Case

It's very common to have a custom domain ( `dev.rds.mydomain.com`) as a CNAME to an RDS cluster (`dev-cluster.cluster-cmdXXXXXX.eu-west-2.rds.amazonaws.com.`)

With the MySQL JDBC driver this Just Works. With the AWS driver it's required to also clusterInstanceHostPattern to inform the driver of :
* RDS account id (`cmdXXXXXX`)
* RDS region (`eu-west-2`)
By adding to the JDBC URL `?clusterInstanceHostPattern=?.cmdXXXXXX.eu-west-2.rds.amazonaws.com`

Which although clearly documented is a barrier to adoption because 'just' swapping in the JDBC URL and .jar does not work as expected.

### Proposed Solution

AWS JDBC driver should be able to look up the current account and region and generate the correct internals details for itself, either from information schemas in the database itself or some other back channel.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### The AWS JDBC Driver for MySQL version used

1.1.0

### JDK version used

openjdk version "11.0.7" 2020-04-14

### Operating System and version

Ubuntu 22.04.1 LTS

**Comment by sergiyvamz** [comment]
Hello @tomchiverton 

Thank you for raising the issue. We understand this issue is causing inconveniences in your workflows. We'll be investigating this issue and will update you when we have a release. We appreciate your patience.

**Comment by tomchiverton** [comment]
So, is this fixed? Or just taking so long the bot has jumped in?

**Comment by aaron-congo** [comment]
Hi @tomchiverton,

Thanks for reaching out with this feature request. I've done some investigation and have some notes/questions. As far as I can tell the only way to auto-acquire the URL attributes you mentioned is by using the AWS SDK [DescribeDBClusters API call](https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_DescribeDBClusters.html). There are some potential issues with doing this:
- to perform the DescribeDBClusters call the user needs to have the proper credentials and access. They will need to either configure their credentials through some properties or rely on their aws credentialsfile if they have one. More info can be found [here](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html#credentials-default). It is unclear whether it is sensible to assume the credentials in the credentials file would be the correct ones to use in the request by default. The credentials the user authorizes with must also be authorized to perform the DescribeDBClusters action on their cluster(s)
- the user will need to either specify the region of their cluster through a configuration property or rely on the default region specified in their aws config file (more details [here](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/setup.html#setting-the-default-aws-region)). Otherwise, the driver doesn't have any way to know which region the desired cluster is in. It is unclear whether it is sensible to assume the region in the config file is also the region of the cluster by default
- given the need to execute the DescribeDBClusters call, there will be a performance penalty when connecting using this method. Whether this penalty would occur only on initial connection or anytime a new connection is established is unclear (requires more investigation)

In summary, this feature request definitely seems useful in theory. However, given the need for the correct AWS credentials and region settings it seems like it may actually be a more complicated user experience then the clusterInstanceHostPattern. Given that the goal of this feature request is to remove configuration parameters but the user may be required to configure their AWS credentials and/or desired region instead, and considering the potential performance penalty, is this still a feature that you would consider useful?

**Comment by rudolfv** [comment]
@aaron-congo This feature would also be very helpful for us. What is the possibility of creating a separate AWS API call that will only provide the URL attributes required? I'm going on the assumption is that this stripped-down API call will also be more performant, but this may incorrect.

**Comment by tomchiverton** [comment]
> given the need for the correct AWS credentials and region settings it seems like it may actually be a more complicated user experience then the clusterInstanceHostPattern. 

All the places we use the driver, have instance roles and access to IMDSv2 so would be be able to figure out region, and make some sort of built-in API call fine. The driver could still support an explicit parameter for user cases where the code is not running somewhere with all the right privileges.

**Comment by davecramer** [comment]
I would think the solution might be to just provide the name to append to the instance id. Thoughts ? You may have to be careful how you name things, but it would be simpler then trying to figure it out. I'm a little worried about the time to execute the API call

**Comment by tomchiverton** [comment]
It'd add a tiny overhead once per server start I'd assume. For us running an app under Tomcat this would be fine. If you are using it with a Java Lambda or similar, which is already slow, that'd be another thing I guess.

**Comment by davecramer** [comment]
@tomchiverton I suppose. Be worth testing I guess.

**Comment by jolexa** [comment]
fwiw, I found that this GH Issue to be more prescriptive than the project's README regarding the value of `clusterInstanceHostPattern` with the given example use case. I wasn't following the `customhost:1234` example very well

**Comment by davecramer** [comment]
Closed with #360
